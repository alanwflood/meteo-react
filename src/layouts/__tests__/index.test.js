import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import App from "../index";
jest.mock("../../components/weather");
// Mock out places autocomplete
jest.mock("react-places-autocomplete", () => {
  const React = require("react"); // eslint-disable-line
  class PlacesAutocomplete extends React.Component {
    renderProps = {
      getInputProps: jest.fn(({ placeholder, className }) => ({
        placeholder,
        className,
      })),
      suggestions: [],
      getSuggestionItemProps: jest.fn(),
    };

    render() {
      // eslint-disable-next-line react/prop-types
      return <>{this.props.children(this.renderProps)}</>;
    }
  }

  return {
    __esModule: true,
    default: PlacesAutocomplete,
    geocodeByAddress: jest
      .fn()
      .mockImplementation(() => Promise.resolve("address")),
    getLatLng: jest.fn().mockImplementation(() => ({ lat: 0, lng: 0 })),
  };
});

describe("Main Layout", () => {
  beforeEach(() => {
    Object.defineProperty(window, "localStorage", {
      value: {
        getItem: jest.fn(() => null),
        setItem: jest.fn(() => null),
      },
      writable: true,
    });
  });

  it("on initial load it shows a logo and searchbar", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("calls localStorage getItem 3 times on render", () => {
    render(<App />);
    expect(window.localStorage.getItem).toHaveBeenCalledTimes(3);
  });

  it("after searching it calls localStorage setItem", async () => {
    window.google = {};
    const { container } = render(<App />);
    const button = container.querySelector("button");
    fireEvent.click(button);
    // Wait for promise to complete
    await waitFor(() => {
      expect(window.localStorage.setItem).toHaveBeenCalledTimes(3);
      expect(container).toMatchSnapshot();
    }, 0);
  });
});
