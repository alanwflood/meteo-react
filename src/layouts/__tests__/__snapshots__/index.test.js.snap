// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Layout after searching it calls localStorage setItem 1`] = `
<div>
  <main
    class="main-content"
  >
    <form
      class="search"
    >
      <div
        class="search-bar"
      >
        <input
          aria-labelledby="searchButton"
          class="search-input"
          placeholder="Search Places ..."
        />
        <div
          class="search-dropdown"
        />
      </div>
      <button
        aria-label="Search for a location to Fetch weather for"
        id="searchButton"
        type="submit"
      >
        <span>
          Search
        </span>
        <test-file-stub />
      </button>
    </form>
  </main>
</div>
`;

exports[`Main Layout on initial load it shows a logo and searchbar 1`] = `
<div>
  <main
    class="main-content"
  >
    <div
      class="logo"
    >
      <h1>
        <div>
          Meteo
        </div>
        <small>
          Weather at a Glance
        </small>
      </h1>
      <div
        class="tagline"
      >
        Search for a location above to get your 5 day weather forecast!
      </div>
    </div>
  </main>
</div>
`;
